### PR TITLE
samples: Bluetooth: Demonstrate MTU update

### PR DESCRIPTION
**Pull request by theob-pro**
Add a new sample and a BabbleSim test for it. The goal of this sample is to demonstrate how to update the MTU to allow larger packet transmission as requested by https://github.com/zephyrproject-rtos/zephyr/issues/25648.

### CODE REVIEW

File: samples/bluetooth/mtu_update/README.rst [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+.. [#] Bluetooth Core Specification v. 5.3: Vol. 3, Part F, 3.4.2
+.. [#] Bluetooth Core Specification v. 5.3: Vol. 3, Part A, 5.1
+.. [#] Bluetooth Core Specification v. 5.3: Vol. 3, Part A, 7.3
+
+Building and Running
+********************
+
+This sample can be found under :zephyr_file:`samples/bluetooth/mtu_update` in
+the Zephyr tree.
```
Comment (jori-nordic) [inline samples/bluetooth/mtu_update/README.rst:65]: can you add a section on the hardware setup, e.g. that you need to flash this part to a device ,and that part to another device. 
and a little text about what you're expected to see if it works.

File: samples/bluetooth/mtu_update/peripheral/prj.conf [hunk 2]
```diff
@@ -0,0 +1,13 @@
+CONFIG_BT=y
+CONFIG_BT_PERIPHERAL=y
+CONFIG_BT_DEVICE_NAME="Zephyr Peripheral MTU Update Sample"
+
+CONFIG_LOG=y
+CONFIG_BT_L2CAP_LOG_LEVEL_DBG=y
+
+# HCI ACL buffers size
+# BT_L2CAP_RX_MTU=CONFIG_BT_BUF_ACL_RX_SIZE-BT_L2CAP_HDR_SIZE
```
Comment (jori-nordic) [inline samples/bluetooth/mtu_update/peripheral/prj.conf:9]: same comment about spaces

File: samples/bluetooth/mtu_update/peripheral/src/peripheral_mtu_update.c [hunk 3]
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
```
Comment (jori-nordic) [inline samples/bluetooth/mtu_update/peripheral/src/peripheral_mtu_update.c:2]: check all copyright notices

**Review by jori-nordic — CHANGES_REQUESTED** [review]

File: samples/bluetooth/mtu_update/README.rst [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+#####################
+
+Q&A:
+****
+
+  Question: What do I have to do to update my ATT MTU?
+
+| Answer: To get an ATT MTU of x octets, do the following:
+| Set :kconfig:option:`CONFIG_BT_L2CAP_TX_MTU` to at least x
+| Set :kconfig:option:`CONFIG_BT_BUF_ACL_RX_SIZE` to at least x + L2CAP header
```
Comment (alwa-nordic) [inline samples/bluetooth/mtu_update/README.rst:13]: What is the actual number for the size of the L2CAP header? I think we should have the number here.

File: samples/bluetooth/mtu_update/README.rst [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  size (4 octets).
+| Set :kconfig:option:`CONFIG_BT_BUF_ACL_RX_SIZE` to at least x + L2CAP header +
+  SDU length field length (6 octets) if using
+  :kconfig:option:`CONFIG_BT_EATT`.
+| Ensure the remote side supports the same MTUs.
+
+  Question: I only want to *send* large packets. I don't need to receive large.
```
Comment (jori-nordic) [inline samples/bluetooth/mtu_update/README.rst:20]: `large ones` or `large packets`

File: samples/bluetooth/mtu_update/README.rst [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+.. figure:: img/mtu.svg
+     :align: center
+     :alt: Diagram of the MTUs and their corresponding Kconfig symbols
+
+     Diagram of the MTUs and their corresponding Kconfig symbols
+
+Hardware Setup
+**************
+
+This sample use two application, two devices need to be setup.
```
Comment (jori-nordic) [inline samples/bluetooth/mtu_update/README.rst:64]: applications

File: samples/bluetooth/mtu_update/README.rst [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+Building and Running
+********************
+
+This sample can be found under :zephyr_file:`samples/bluetooth/mtu_update` in
+the Zephyr tree.
+
+See :ref:`bluetooth samples section <bluetooth-samples>` for details.
+
+If the devices are close enough, the central should connect to the peripheral
+and send his MTU to the other device. If the MTU exchange succeed, the central
```
Comment (jori-nordic) [inline samples/bluetooth/mtu_update/README.rst:79]: succeeds

File: samples/bluetooth/mtu_update/README.rst [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+should subscribe and then the peripheral will send a large notification. Right
+after receiving the notification the central should unsubscribe.
+
+Here are the output you should have on the devices:
```
Comment (jori-nordic) [inline samples/bluetooth/mtu_update/README.rst:83]: `Here is the output` or `Here are the outputs`.

File: samples/bluetooth/mtu_update/central/src/central_mtu_update.c [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+	if (default_conn) {
+		return;
+	}
+
+	/* We're only interested in connectable events */
+	if (type != BT_GAP_ADV_TYPE_ADV_IND &&
+	    type != BT_GAP_ADV_TYPE_ADV_DIRECT_IND) {
+		return;
+	}
+
```
Comment (alwa-nordic) [inline samples/bluetooth/mtu_update/central/src/central_mtu_update.c:128]: Should we look in the ad data for the service before connecting?

**Review by jori-nordic — APPROVED** [review]

**Review by alwa-nordic — APPROVED** [review]

**Comment by wiba-nordic** [comment]
I'm going over the Zephyr ⁠[sample definition](https://builds.zephyrproject.io/zephyr/pr/53593/docs/samples/sample_definition_and_criteria.html), and point 6 says that a sample should have minimal comments and not be explaining the WHY of something. But the sample currently has quite a bit of them, going into a lot the why and details of an MTU update. Would some of that be better served in a different part of the documentation instead of as part of a sample?

File: samples/bluetooth/mtu_update/README.rst [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+     Diagram of the MTUs and their corresponding Kconfig symbols
+
+Hardware Setup
+**************
+
+This sample use two applications, two devices need to be setup.
+The first one should be flashed with the central and the second one with the
+peripheral.
+
+The two devices need to be close enough to be able to connect.
```
Comment (alwa-nordic) [inline samples/bluetooth/mtu_update/README.rst:69]: It's a bit hard to understand what is meant by this sentence. I think we should reword so it's clear it's a feature.

```suggestion
The two devices will connect only if they are close to each other, thanks to RSSI filtering.
```

**Comment by theob-pro** [comment]
> I'm going over the Zephyr ⁠[sample definition](https://builds.zephyrproject.io/zephyr/pr/53593/docs/samples/sample_definition_and_criteria.html), and point 6 says that a sample should have minimal comments and not be explaining the WHY of something. But the sample currently has quite a bit of them, going into a lot the why and details of an MTU update. Would some of that be better served in a different part of the documentation instead of as part of a sample?

I think that the point 6 refer to code comment and not the README documentation. Anyway, I don't think that limiting the comment on code make a lot of sense either, is there a good reason for that? @carlescufi 
For the "why", you're right, I will remove the sentence citing reasons to update the MTU.

**Review by jhedberg — APPROVED** [review]

**Comment by carlescufi** [comment]
> > I'm going over the Zephyr ⁠[sample definition](https://builds.zephyrproject.io/zephyr/pr/53593/docs/samples/sample_definition_and_criteria.html), and point 6 says that a sample should have minimal comments and not be explaining the WHY of something. But the sample currently has quite a bit of them, going into a lot the why and details of an MTU update. Would some of that be better served in a different part of the documentation instead of as part of a sample?
> 
> I think that the point 6 refer to code comment and not the README documentation. Anyway, I don't think that limiting the comment on code make a lot of sense either, is there a good reason for that? @carlescufi For the "why", you're right, I will remove the sentence citing reasons to update the MTU.

In this particular case the documentation included in the sample is definitely helpful and belongs in it.

**Pull request merged**